// @fukutotojido/z-engine/dist/index.js
//#region node_modules/@fukutotojido/z-engine/dist/index.js
var __commonJS = (cb, mod) => () => (mod || cb((mod = { exports: {} }).exports, mod), mod.exports);
var require_scanner = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.Scanner = void 0;
	function testWithString(target, pattern) {
		return target.startsWith(pattern) ? [pattern] : [];
	}
	function testWithRegExp(target, pattern) {
		const result = pattern.exec(target);
		if (!result) return [];
		return result;
	}
	class Scanner {
		constructor(target, trim = true) {
			this.target = target;
			if (trim) this.trimWhitespaces();
		}
		scan(pattern, trim = true) {
			const [matched, ...groups] = typeof pattern === "string" ? testWithString(this.target, pattern) : testWithRegExp(this.target, pattern);
			if (!matched) return null;
			this.target = this.target.substring(matched.length);
			if (trim) this.trimWhitespaces();
			return [matched, ...groups];
		}
		trimWhitespaces() {
			this.target = this.target.replace(/^\s+/, "");
		}
	}
	exports.Scanner = Scanner;
});
var require_stmt = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.Statements = exports.isExprStatement = exports.ExprStatement = exports.DefFuncStatement = void 0;
	class DefFuncStatement {
		constructor(name) {
			this.name = name;
		}
		dump() {
			return { deffun: this.name };
		}
	}
	exports.DefFuncStatement = DefFuncStatement;
	class ExprStatement {
		constructor(expr) {
			this.expr = expr;
		}
		dump() {
			return { expr: this.expr.dump() };
		}
	}
	exports.ExprStatement = ExprStatement;
	function isExprStatement(stmt) {
		return "expr" in stmt;
	}
	exports.isExprStatement = isExprStatement;
	class Statements {
		constructor() {
			this.stmts = [];
		}
		push(stmt) {
			this.stmts.push(stmt);
		}
		dump() {
			return { statements: this.stmts.map((stmt) => stmt.dump()) };
		}
	}
	exports.Statements = Statements;
});
var require_expr = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.OptionalExpr = exports.FilteredExpr = exports.ParallelExpr = exports.BinaryOperatorExpr = exports.IndexedExpr = exports.SpreadIndexer = exports.KeyIndexer = exports.RangeIndexer = exports.ObjectExpr = exports.Field = exports.ArrayExpr = exports.isIdentityExpr = exports.IdentityExpr = exports.RecursiveDescendantExpr = exports.IdExpr = exports.FunctionCallExpr = exports.StringLiteralExpr = exports.LiteralExpr = void 0;
	class LiteralExpr {
		constructor(value) {
			this.value = value;
		}
		dump() {
			return this.value;
		}
	}
	exports.LiteralExpr = LiteralExpr;
	class StringLiteralExpr {
		constructor(parts) {
			this.parts = parts;
		}
		get value() {
			return this.parts.join();
		}
		dump() {
			return this.value;
		}
	}
	exports.StringLiteralExpr = StringLiteralExpr;
	class FunctionCallExpr {
		constructor(funcname, args) {
			this.funcname = funcname;
			this.args = args;
		}
		dump() {
			return { call: { [this.funcname]: this.args } };
		}
	}
	exports.FunctionCallExpr = FunctionCallExpr;
	class IdExpr {
		constructor(value) {
			this.value = value;
		}
		dump() {
			return { id: this.value };
		}
	}
	exports.IdExpr = IdExpr;
	class RecursiveDescendantExpr {
		dump() {
			return { recursiveDescendant: 0 };
		}
	}
	exports.RecursiveDescendantExpr = RecursiveDescendantExpr;
	class IdentityExpr {
		constructor() {
			this.identity = true;
		}
		dump() {
			return { identity: 0 };
		}
	}
	exports.IdentityExpr = IdentityExpr;
	function isIdentityExpr(expr) {
		return "identity" in expr;
	}
	exports.isIdentityExpr = isIdentityExpr;
	class ArrayExpr {
		constructor() {
			this.elements = [];
		}
		push(expr) {
			this.elements.push(expr);
		}
		dump() {
			return { array: this.elements.map((x) => x.dump()) };
		}
	}
	exports.ArrayExpr = ArrayExpr;
	class Field {
		constructor(key, value = null) {
			this.key = key;
			this.value = value;
		}
		dump() {
			var _a;
			return {
				key: this.key.dump(),
				value: (_a = this.value) === null || _a === void 0 ? void 0 : _a.dump()
			};
		}
	}
	exports.Field = Field;
	class ObjectExpr {
		constructor() {
			this.fields = [];
		}
		push(field) {
			this.fields.push(field);
		}
		dump() {
			return { object: this.fields.map((x) => x.dump()) };
		}
	}
	exports.ObjectExpr = ObjectExpr;
	class RangeIndexer {
		constructor(first, last = null) {
			this.first = first;
			this.last = last;
		}
		dump() {
			if (this.first) if (this.last) return [this.first.dump(), this.last.dump()];
			else return [this.first.dump(), "last"];
			else if (this.last) return ["start", this.last.dump()];
		}
	}
	exports.RangeIndexer = RangeIndexer;
	class KeyIndexer {
		constructor(key) {
			this.key = key;
		}
		dump() {
			return { key: this.key };
		}
	}
	exports.KeyIndexer = KeyIndexer;
	class SpreadIndexer {
		dump() {
			return { spread: 0 };
		}
	}
	exports.SpreadIndexer = SpreadIndexer;
	class IndexedExpr {
		constructor(expr, indexer, isOptional) {
			this.expr = expr;
			this.indexer = indexer;
			this.isOptional = isOptional;
		}
		dump() {
			return {
				indexing: this.expr.dump(),
				indexer: this.indexer.dump(),
				isOptional: this.isOptional
			};
		}
	}
	exports.IndexedExpr = IndexedExpr;
	class BinaryOperatorExpr {
		constructor(op, lhs, rhs) {
			this.op = op;
			this.lhs = lhs;
			this.rhs = rhs;
		}
		dump() {
			var _a, _b;
			return { [this.op]: [(_a = this.lhs) === null || _a === void 0 ? void 0 : _a.dump(), (_b = this.rhs) === null || _b === void 0 ? void 0 : _b.dump()] };
		}
	}
	exports.BinaryOperatorExpr = BinaryOperatorExpr;
	class ParallelExpr {
		constructor() {
			this.exprs = [];
		}
		push(expr) {
			this.exprs.push(expr);
		}
		dump() {
			var _a;
			if (this.exprs.length < 2) return (_a = this.exprs[0]) === null || _a === void 0 ? void 0 : _a.dump();
			return { parallel: this.exprs.map((x) => x.dump()) };
		}
	}
	exports.ParallelExpr = ParallelExpr;
	class FilteredExpr {
		constructor() {
			this.exprs = [];
		}
		push(expr) {
			this.exprs.push(expr);
		}
		dump() {
			var _a;
			if (this.exprs.length < 2) return (_a = this.exprs[0]) === null || _a === void 0 ? void 0 : _a.dump();
			return { filtered: this.exprs.map((x) => x.dump()) };
		}
	}
	exports.FilteredExpr = FilteredExpr;
	class OptionalExpr {
		constructor(expr) {
			this.expr = expr;
		}
		dump() {
			optional: this.expr.dump();
		}
	}
	exports.OptionalExpr = OptionalExpr;
});
var require_parser = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.parse = void 0;
	var stmt_1 = require_stmt();
	var expr_1 = require_expr();
	class ParseError extends Error {}
	function parseFunctionDefinition(scanner) {
		if (scanner.target.length > 0) {
			if (scanner.scan("def")) throw new ParseError(`not supported: user-defined funtions`);
		}
		return null;
	}
	function parseNumberLiteral(scanner) {
		const result = scanner.scan(/^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?/);
		if (result) return new expr_1.LiteralExpr(parseFloat(result[0]));
		else return null;
	}
	function parseBooleanLiteral(scanner) {
		if (scanner.scan("true")) return new expr_1.LiteralExpr(true);
		if (scanner.scan("false")) return new expr_1.LiteralExpr(false);
		return null;
	}
	function parseStringLiteral(scanner) {
		var _a, _b;
		const [quote] = (_a = scanner.scan(/^('|")/, false)) !== null && _a !== void 0 ? _a : [];
		if (!quote) return null;
		const parts = [];
		const pattern = new RegExp(`^[^${quote}\\\\]*`);
		while (scanner.target.length > 0) {
			const [part] = (_b = scanner.scan(pattern, false)) !== null && _b !== void 0 ? _b : [];
			if (part) parts.push(part);
			if (scanner.scan(quote)) return new expr_1.StringLiteralExpr(parts);
			if (scanner.scan("\\", false)) throw new ParseError(`escape sequence is not supported yet: ${scanner.target}`);
		}
		throw new ParseError(`cannot find closing ${quote}`);
	}
	function parseIdentifier(scanner) {
		var _a;
		const [id] = (_a = scanner.scan(/^[a-zA-Z_][a-zA-Z_0-9]*/)) !== null && _a !== void 0 ? _a : [];
		if (id) return new expr_1.IdExpr(id);
		return null;
	}
	function parseDoulbeDot(scanner) {
		if (scanner.scan("..")) return new expr_1.RecursiveDescendantExpr();
		return null;
	}
	function parseDot(scanner) {
		if (scanner.scan(".")) return new expr_1.IdentityExpr();
		return null;
	}
	function parseArrayElement(scanner) {
		const first = parseStandaloneExpression(scanner);
		if (!first) return null;
		const root = new expr_1.FilteredExpr();
		root.push(first);
		while (scanner.target.length > 0) {
			if (!scanner.scan("|")) break;
			const next = parseStandaloneExpression(scanner);
			if (!next) throw new ParseError(`missing right side expression right of '|': ${scanner.target}`);
			root.push(next);
		}
		return root;
	}
	function parseArrayExpr(scanner) {
		if (!scanner.scan("[")) return null;
		const expr = new expr_1.ArrayExpr();
		const first = parseArrayElement(scanner);
		if (!first) {
			if (scanner.scan("]")) return expr;
			throw new ParseError(`missing closing bracket: ${scanner.target}`);
		}
		expr.push(first);
		while (scanner.target.length > 0) {
			if (scanner.scan("]")) return expr;
			if (!scanner.scan(",")) throw new ParseError(`missing comma in array: ${scanner.target}`);
			const elem = parseArrayElement(scanner);
			if (!elem) throw new ParseError(`missing closing bracket: ${scanner.target}`);
			expr.push(elem);
		}
		if (!scanner.scan("]")) throw new ParseError(`missing closing bracket: ${scanner.target}`);
		return expr;
	}
	function parseObjectFieldKey(scanner) {
		const grp = parseGroupExpr(scanner);
		if (grp) return grp;
		const str = parseStringLiteral(scanner);
		if (str) return str;
		const id = parseIdentifier(scanner);
		if (id) return id;
		return null;
	}
	function parseObjectField(scanner) {
		const key = parseObjectFieldKey(scanner);
		if (!key) throw new ParseError(`missing field key: ${scanner.target}`);
		if (!scanner.scan(":")) return new expr_1.Field(key);
		const value = parseFilteredExpression(scanner, false);
		if (!value) throw new ParseError(`missing field value: ${scanner.target}`);
		return new expr_1.Field(key, value);
	}
	function parseObjectExpr(scanner) {
		if (!scanner.scan("{")) return null;
		const expr = new expr_1.ObjectExpr();
		if (scanner.scan("}")) return expr;
		const first = parseObjectField(scanner);
		if (!first) throw new ParseError(`missing closing brace: ${scanner.target}`);
		expr.push(first);
		while (scanner.target.length > 0) {
			if (scanner.scan("}")) return expr;
			if (!scanner.scan(",")) throw new ParseError(`missing comma in object: ${scanner.target}`);
			const field = parseObjectField(scanner);
			if (!field) throw new ParseError(`missing closing brace: ${scanner.target}`);
			expr.push(field);
		}
		throw new ParseError(`missing closing brace: ${scanner.target}`);
	}
	function parseGroupExpr(scanner) {
		if (!scanner.scan("(")) return null;
		const expr = parseExpression(scanner);
		if (!scanner.scan(")")) throw new ParseError(`missing closing paren: ${scanner.target}`);
		return expr;
	}
	function parseFunctionCall(scanner) {
		const id = parseIdentifier(scanner);
		if (!id) return null;
		const args = [];
		if (scanner.scan("(")) {
			for (;;) {
				const arg = parseExpression(scanner);
				if (!arg) break;
				args.push(arg);
				if (scanner.scan(";")) continue;
			}
			if (!scanner.scan(")")) throw new ParseError(`missing closing paren: ${scanner.target}`);
		}
		return new expr_1.FunctionCallExpr(id.value, args);
	}
	function parsePrimitiveExpression(scanner) {
		const grp = parseGroupExpr(scanner);
		if (grp) return grp;
		const ary = parseArrayExpr(scanner);
		if (ary) return ary;
		const obj = parseObjectExpr(scanner);
		if (obj) return obj;
		const str = parseStringLiteral(scanner);
		if (str) return str;
		const bool = parseBooleanLiteral(scanner);
		if (bool) return bool;
		const num = parseNumberLiteral(scanner);
		if (num) return num;
		const call = parseFunctionCall(scanner);
		if (call) {
			if (call.funcname == "null") return new expr_1.LiteralExpr(null);
			return call;
		}
		const recursive = parseDoulbeDot(scanner);
		if (recursive) return recursive;
		const dot = parseDot(scanner);
		if (dot) return dot;
		return null;
	}
	function parseArrayLikeIndexer(scanner) {
		if (scanner.target.length == 0) return null;
		if (!scanner.scan("[")) return null;
		const start = parseExpression(scanner);
		const isRange = scanner.scan(":");
		const last = isRange ? parseExpression(scanner) : null;
		if (!scanner.scan("]")) throw new ParseError(`no closing bracket: ${scanner.target}`);
		if (isRange) return new expr_1.RangeIndexer(start, last);
		else if (start === null) return new expr_1.SpreadIndexer();
		else return new expr_1.KeyIndexer(start);
	}
	function parseDotIndexer(scanner) {
		if (scanner.target.length == 0) return null;
		const str = parseStringLiteral(scanner);
		if (str) return new expr_1.KeyIndexer(str);
		const id = parseIdentifier(scanner);
		if (id) return new expr_1.KeyIndexer(id);
		return null;
	}
	function parseIndexer(scanner) {
		if (scanner.target.length > 0) {
			const arrayIndexer = parseArrayLikeIndexer(scanner);
			if (arrayIndexer) return arrayIndexer;
			if (!scanner.scan(".")) return null;
			const dotIndexer = parseDotIndexer(scanner);
			if (dotIndexer) return dotIndexer;
			throw new ParseError(`no valid key: ${scanner.target}`);
		}
		return null;
	}
	function parseIndexedExpr(scanner) {
		if (scanner.target.length > 0) {
			let lhs = parsePrimitiveExpression(scanner);
			if (!lhs) return null;
			if (expr_1.isIdentityExpr(lhs)) {
				const indexer = parseDotIndexer(scanner);
				if (indexer) lhs = new expr_1.IndexedExpr(lhs, indexer, false);
			}
			if (scanner.scan("?")) lhs = new expr_1.OptionalExpr(lhs);
			while (scanner.target.length > 0) {
				const indexer = parseIndexer(scanner);
				if (!indexer) break;
				lhs = new expr_1.IndexedExpr(lhs, indexer, false);
				if (scanner.scan("?")) lhs = new expr_1.OptionalExpr(lhs);
			}
			return lhs;
		}
		return null;
	}
	function parseMultiplicativeExpr(scanner) {
		let lhs = null;
		let op = null;
		while (scanner.target.length > 0) {
			const subexpr = parseIndexedExpr(scanner);
			if (!subexpr) break;
			if (lhs && op) {
				lhs = new expr_1.BinaryOperatorExpr(op, lhs, subexpr);
				op = null;
			} else lhs = subexpr;
			const result = scanner.scan(/^(\*|\/(?!\/)|%)/);
			if (!result) break;
			op = result[0];
		}
		return lhs;
	}
	function parseAdditiveExpr(scanner) {
		let lhs = null;
		let op = null;
		while (scanner.target.length > 0) {
			const subexpr = parseMultiplicativeExpr(scanner);
			if (!subexpr) break;
			if (lhs && op) {
				lhs = new expr_1.BinaryOperatorExpr(op, lhs, subexpr);
				op = null;
			} else lhs = subexpr;
			const result = scanner.scan(/^(\+|\-)/);
			if (!result) break;
			op = result[0];
		}
		return lhs;
	}
	function parseComparativeExpr(scanner) {
		let lhs = null;
		let op = null;
		while (scanner.target.length > 0) {
			const subexpr = parseAdditiveExpr(scanner);
			if (!subexpr) break;
			if (lhs && op) {
				lhs = new expr_1.BinaryOperatorExpr(op, lhs, subexpr);
				op = null;
			} else lhs = subexpr;
			const result = scanner.scan(/^(==|!=|<=|<|>=|>)/);
			if (!result) break;
			op = result[0];
		}
		return lhs;
	}
	function parseLogicalAndExpr(scanner) {
		let lhs = null;
		while (scanner.target.length > 0) {
			const subexpr = parseComparativeExpr(scanner);
			if (!subexpr) break;
			if (lhs) lhs = new expr_1.BinaryOperatorExpr("and", lhs, subexpr);
			else lhs = subexpr;
			if (!scanner.scan("and")) break;
		}
		return lhs;
	}
	function parseLogicalOrExpression(scanner) {
		let lhs = null;
		while (scanner.target.length > 0) {
			const subexpr = parseLogicalAndExpr(scanner);
			if (!subexpr) break;
			if (lhs) lhs = new expr_1.BinaryOperatorExpr("or", lhs, subexpr);
			else lhs = subexpr;
			if (!scanner.scan("or")) break;
		}
		return lhs;
	}
	function parseArithmeticExpression(scanner) {
		return parseLogicalOrExpression(scanner);
	}
	function parseAlternativeExpression(scanner) {
		let lhs = null;
		while (scanner.target.length > 0) {
			const subexpr = parseArithmeticExpression(scanner);
			if (!subexpr) break;
			if (lhs) lhs = new expr_1.BinaryOperatorExpr("//", lhs, subexpr);
			else lhs = subexpr;
			if (!scanner.scan("//")) break;
		}
		return lhs;
	}
	function parseStandaloneExpression(scanner) {
		return parseAlternativeExpression(scanner);
	}
	function parseParallelExpression(scanner) {
		const first = parseStandaloneExpression(scanner);
		if (!first) return null;
		const root = new expr_1.ParallelExpr();
		root.push(first);
		while (scanner.target.length > 0) {
			if (!scanner.scan(",")) break;
			const next = parseStandaloneExpression(scanner);
			if (!next) throw new ParseError(`missing expr right side of ',': ${scanner.target}`);
			root.push(next);
		}
		return root;
	}
	function parseFilteredExpression(scanner, allowParallel = true) {
		const first = allowParallel ? parseParallelExpression(scanner) : parseStandaloneExpression(scanner);
		if (!first) return null;
		const root = new expr_1.FilteredExpr();
		root.push(first);
		while (scanner.target.length > 0) {
			if (!scanner.scan("|")) break;
			const next = allowParallel ? parseParallelExpression(scanner) : parseStandaloneExpression(scanner);
			if (!next) throw new ParseError(`missing expr right side of '|': ${scanner.target}`);
			root.push(next);
		}
		return root;
	}
	function parseExpression(scanner) {
		return parseFilteredExpression(scanner);
	}
	function parseStatement(scanner) {
		if (scanner.target.length > 0) {
			const deffun = parseFunctionDefinition(scanner);
			if (deffun) return deffun;
			const expr = parseExpression(scanner);
			if (expr) return new stmt_1.ExprStatement(expr);
		}
		return null;
	}
	function parseStatements(scanner) {
		const stmts = new stmt_1.Statements();
		while (scanner.target.length > 0) {
			const stmt = parseStatement(scanner);
			if (!stmt) break;
			stmts.push(stmt);
			if (!scanner.scan(";")) break;
		}
		if (scanner.target.length > 0) throw new ParseError(`cannot parse whole pattern: consumed = ${JSON.stringify(stmts.dump())} rest = <${scanner.target}>`);
		return stmts;
	}
	function parse(scanner) {
		return parseStatements(scanner);
	}
	exports.parse = parse;
});
var require_context = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.ContextWithTarget = exports.Context = exports.isJsonMap = void 0;
	function isJsonMap(x) {
		if (x === null) return false;
		if (typeof x === "number" || typeof x === "string" || typeof x === "boolean") return false;
		if (Array.isArray(x)) return false;
		return true;
	}
	exports.isJsonMap = isJsonMap;
	class Context {
		constructor(...values) {
			this.values = values;
		}
		withTarget(target) {
			return new ContextWithTarget(target, ...this.values);
		}
	}
	exports.Context = Context;
	class ContextWithTarget extends Context {
		constructor(target, ...values) {
			super(...values);
			this.target = target;
		}
	}
	exports.ContextWithTarget = ContextWithTarget;
});
var require_error = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.UnexpectedIdError = exports.RuntimeError = void 0;
	class RuntimeError extends Error {}
	exports.RuntimeError = RuntimeError;
	class UnexpectedIdError extends RuntimeError {
		constructor(id) {
			super(`unexpected identifier: ${id}`);
		}
	}
	exports.UnexpectedIdError = UnexpectedIdError;
});
var require_ensure = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.ensureSlicable = exports.ensureArrayIndexable = exports.ensureObjectKeyOrArrayIndexable = exports.ensureObjectKey = exports.ensureArray = exports.ensureNumberValue = exports.ensureValue = void 0;
	var error_1 = require_error();
	function ensureValue(value) {
		if ("error" in value) throw value.error;
		if ("id" in value) throw new error_1.UnexpectedIdError(value.id);
		return value.value;
	}
	exports.ensureValue = ensureValue;
	function ensureNumberValue(value) {
		if ("error" in value) throw value.error;
		if ("id" in value) throw new error_1.UnexpectedIdError(value.id);
		if (typeof value.value !== "number") throw new error_1.RuntimeError(`unexpected non-number value: ${value.value}`);
		return value.value;
	}
	exports.ensureNumberValue = ensureNumberValue;
	function ensureArray(value) {
		const values = ensureValue(value);
		if (Array.isArray(values)) return values;
		throw new error_1.RuntimeError(`unexpected non-array value: ${values}`);
	}
	exports.ensureArray = ensureArray;
	function ensureObjectKey(value) {
		if ("error" in value) throw value.error;
		if ("id" in value) return value.id;
		if (typeof value.value === "string") return value.value;
		throw new error_1.RuntimeError(`unexpected non-string value for object key: ${value.value}`);
	}
	exports.ensureObjectKey = ensureObjectKey;
	function ensureObjectKeyOrArrayIndexable(value) {
		if ("error" in value) throw value.error;
		if ("id" in value) return value.id;
		if (typeof value.value === "string" || typeof value.value === "number") return value.value;
		throw new error_1.RuntimeError(`unexpected non-string value for object key: ${value.value}`);
	}
	exports.ensureObjectKeyOrArrayIndexable = ensureObjectKeyOrArrayIndexable;
	function ensureArrayIndexable(value, size) {
		if ("error" in value) throw value.error;
		if ("id" in value) throw new error_1.RuntimeError(`unexpected identifier for array index: ${value.id}`);
		if (typeof value.value === "number") return value.value < 0 ? size + value.value : value.value;
		throw new error_1.RuntimeError(`unexpected non-string value for array index: ${value.value}`);
	}
	exports.ensureArrayIndexable = ensureArrayIndexable;
	function ensureSlicable(value) {
		if ("error" in value) throw value.error;
		if ("value" in value) {
			if (typeof value.value === "string" || Array.isArray(value.value)) return value.value;
			throw new error_1.RuntimeError(`cannot make slice from non-array value: ${value.value}`);
		}
		throw new error_1.RuntimeError(`cannot make slice from identifier: ${value.id}`);
	}
	exports.ensureSlicable = ensureSlicable;
});
var require_utility = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.mergeObject = exports.objectCompare = exports.arrayCompare = exports.valueCompare = exports.isNullish = exports.isFalsy = exports.flatten = exports.expandCombination = void 0;
	var context_1 = require_context();
	function expandCombination(xs, initValue, expand) {
		if (xs.length === 0) return [];
		return xs.reduce((prevValues, x) => {
			return prevValues.reduce((a, prevValue) => [...a, ...expand(prevValue, x)], []);
		}, [initValue]);
	}
	exports.expandCombination = expandCombination;
	function flatten(xs) {
		return xs.reduce((acc, x) => [...acc, ...x], []);
	}
	exports.flatten = flatten;
	function isFalsy(value) {
		return value === void 0 || value === null || value === false;
	}
	exports.isFalsy = isFalsy;
	function isNullish(value) {
		return value === void 0 || value === null;
	}
	exports.isNullish = isNullish;
	function valueCompare(lhs, rhs) {
		if (lhs === null) return rhs === null ? 0 : -1;
		if (rhs === null) return 1;
		if (lhs === false) return rhs === false ? 0 : -1;
		if (rhs === false) return 1;
		if (lhs === true) return rhs === true ? 0 : -1;
		if (rhs === true) return 1;
		if (typeof lhs === "number") {
			if (typeof rhs === "number") return lhs < rhs ? -1 : lhs > rhs ? 1 : 0;
			return -1;
		}
		if (typeof rhs === "number") return 1;
		if (typeof lhs === "string") {
			if (typeof rhs === "string") return lhs < rhs ? -1 : lhs > rhs ? 1 : 0;
			return -1;
		}
		if (typeof rhs === "string") return 1;
		if (Array.isArray(lhs)) {
			if (Array.isArray(rhs)) return arrayCompare(lhs, rhs);
			return -1;
		}
		if (Array.isArray(rhs)) return 1;
		return objectCompare(lhs, rhs);
	}
	exports.valueCompare = valueCompare;
	function arrayCompare(lhs, rhs) {
		for (let i = 0;; ++i) if (i < lhs.length) if (i < rhs.length) {
			const r = valueCompare(lhs[i], rhs[i]);
			if (r !== 0) return r;
		} else return -1;
		else if (i < rhs.length) return 1;
		else return 0;
	}
	exports.arrayCompare = arrayCompare;
	function objectCompare(lhs, rhs) {
		const r = arrayCompare(Object.keys(lhs), Object.keys(rhs));
		if (r !== 0) return r;
		return arrayCompare(Object.values(lhs), Object.values(rhs));
	}
	exports.objectCompare = objectCompare;
	function mergeObjects(...xs) {
		return xs.reduce((acc, x) => Object.assign(Object.assign({}, acc), x), {});
	}
	function mergeObject(lhs, rhs, overwrite) {
		if (overwrite) return mergeObjects(lhs, rhs);
		const tmp = Object.assign({}, lhs);
		Object.keys(rhs).forEach((key) => {
			const r = rhs[key];
			if (key in lhs) {
				const l = lhs[key];
				if (context_1.isJsonMap(l) && context_1.isJsonMap(r)) tmp[key] = mergeObject(l, r, false);
				else tmp[key] = r;
			} else tmp[key] = r;
		});
		return tmp;
	}
	exports.mergeObject = mergeObject;
});
var require_operators = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.equalOperator = exports.notEqualOperator = exports.lessThanOrEqualToOperator = exports.lessThanOperator = exports.greaterThanOrEqualToOperator = exports.greaterThanOperator = exports.alternativeOperator = exports.orOperator = exports.andOperator = exports.minusOperator = exports.addOperator = exports.moduloOperator = exports.divideOperator = exports.multiplyOperator = void 0;
	var context_1 = require_context();
	var utility_1 = require_utility();
	var error_1 = require_error();
	function multiplyOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") return lhs * rhs;
		else if (typeof lhs == "string" && typeof rhs == "number") {
			if (rhs == 0) return null;
			return [...Array(rhs)].map(() => lhs).join();
		} else if (typeof rhs == "string" && typeof lhs == "number") {
			if (lhs == 0) return null;
			return [...Array(lhs)].map(() => rhs).join();
		} else if (context_1.isJsonMap(lhs) && context_1.isJsonMap(rhs)) return utility_1.mergeObject(lhs, rhs, false);
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.multiplyOperator = multiplyOperator;
	function divideOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") {
			if (rhs === 0) throw new error_1.RuntimeError(`divison by zero`);
			return lhs / rhs;
		} else if (typeof lhs == "string" && typeof rhs == "string") return lhs.split(rhs);
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.divideOperator = divideOperator;
	function moduloOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") return lhs % rhs;
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.moduloOperator = moduloOperator;
	function addOperator(lhs, rhs) {
		if (lhs === null) return rhs;
		else if (rhs === null) return lhs;
		else if (typeof lhs == "number" && typeof rhs == "number") return lhs + rhs;
		else if (typeof lhs == "string" && typeof rhs == "string") return lhs + rhs;
		else if (Array.isArray(lhs) && Array.isArray(rhs)) return [...lhs, ...rhs];
		else if (context_1.isJsonMap(lhs) && context_1.isJsonMap(rhs)) return utility_1.mergeObject(lhs, rhs, true);
		throw new error_1.RuntimeError(`cannot apply binary operator '+' with ${JSON.stringify(lhs)} and ${JSON.stringify(rhs)}`);
	}
	exports.addOperator = addOperator;
	function minusOperator(lhs, rhs) {
		if (lhs === null) return rhs;
		else if (rhs === null) return lhs;
		else if (typeof lhs == "number" && typeof rhs == "number") return lhs - rhs;
		else if (Array.isArray(lhs) && Array.isArray(rhs)) return lhs.filter((e) => !rhs.includes(e));
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.minusOperator = minusOperator;
	function andOperator(lhs, rhs) {
		return !utility_1.isFalsy(lhs) && !utility_1.isFalsy(rhs);
	}
	exports.andOperator = andOperator;
	function orOperator(lhs, rhs) {
		return !utility_1.isFalsy(lhs) || !utility_1.isFalsy(rhs);
	}
	exports.orOperator = orOperator;
	function alternativeOperator(lhs, rhs) {
		return !utility_1.isNullish(lhs) ? lhs : rhs;
	}
	exports.alternativeOperator = alternativeOperator;
	function greaterThanOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") return lhs > rhs;
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.greaterThanOperator = greaterThanOperator;
	function greaterThanOrEqualToOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") return lhs >= rhs;
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.greaterThanOrEqualToOperator = greaterThanOrEqualToOperator;
	function lessThanOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") return lhs < rhs;
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.lessThanOperator = lessThanOperator;
	function lessThanOrEqualToOperator(lhs, rhs) {
		if (typeof lhs == "number" && typeof rhs == "number") return lhs <= rhs;
		throw new error_1.RuntimeError(`cannot apply binary operator '*' with ${typeof lhs} and ${typeof rhs}`);
	}
	exports.lessThanOrEqualToOperator = lessThanOrEqualToOperator;
	function notEqualOperator(lhs, rhs) {
		return JSON.stringify(lhs) != JSON.stringify(rhs);
	}
	exports.notEqualOperator = notEqualOperator;
	function equalOperator(lhs, rhs) {
		return JSON.stringify(lhs) == JSON.stringify(rhs);
	}
	exports.equalOperator = equalOperator;
});
var require_predefined = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.combinations = exports.endswith = exports.startswith = exports.rindex = exports.index = exports.indices = exports.contains = exports.reverse = exports.uniqueBy = exports.unique = exports.maxIndex = exports.minIndex = exports.groupBy = exports.sortBy = exports.sort = exports.isInfinite = exports._type = exports.toString = exports.toNumber = exports.range = exports.flattenArray = exports.allOf = exports.anyOf = exports.join = exports.isString = exports.isNormal = exports.isNumber = exports.isBoolean = exports.isIterable = exports.isObject = exports.isArray = exports.fromEntries = exports.toEntries = exports._in = exports.has = exports.keysUnsorted = exports.keys = exports.utf8bytelength = exports.length = void 0;
	var context_1 = require_context();
	var error_1 = require_error();
	var operators_1 = require_operators();
	var utility_1 = require_utility();
	function length(value) {
		if (context_1.isJsonMap(value)) return Object.keys(value).length;
		if (Array.isArray(value)) return value.length;
		if (typeof value === "string") return value.length;
		if (value === null) return 0;
		throw new error_1.RuntimeError(`cannot get length of ${value}`);
	}
	exports.length = length;
	function utf8bytelength(value) {
		if (typeof value !== "string") throw new error_1.RuntimeError(`cannot get byte length of ${value}`);
		return new TextEncoder().encode(value).length;
	}
	exports.utf8bytelength = utf8bytelength;
	function keys(value) {
		if (Array.isArray(value)) return [...value.keys()];
		if (context_1.isJsonMap(value)) return Object.keys(value).sort();
		throw new error_1.RuntimeError(`cannot get keys from: ${value}`);
	}
	exports.keys = keys;
	function keysUnsorted(value) {
		if (Array.isArray(value)) return [...value.keys()];
		if (context_1.isJsonMap(value)) return Object.keys(value);
		throw new error_1.RuntimeError(`cannot get keys from: ${value}`);
	}
	exports.keysUnsorted = keysUnsorted;
	function hasKey(input, key) {
		if (Array.isArray(input)) {
			if (typeof key !== "number") throw new error_1.RuntimeError(`cannot call has(key) with number`);
			return key < input.length;
		}
		if (context_1.isJsonMap(input)) {
			if (typeof key !== "string") throw new error_1.RuntimeError(`cannot call has(key) with non-string: ${JSON.stringify(key)}`);
			return key in input;
		}
		throw new error_1.RuntimeError(`cannot call has(key) against ${input}`);
	}
	function has(input, key) {
		if (key === void 0) throw new error_1.RuntimeError(`missing argment 'key' for has(key)`);
		return hasKey(input, key);
	}
	exports.has = has;
	function _in(input, obj) {
		if (obj === void 0) throw new error_1.RuntimeError(`missing argment 'obj' for in(obj)`);
		return hasKey(obj, input);
	}
	exports._in = _in;
	function toEntries(input) {
		if (!context_1.isJsonMap(input)) throw new error_1.RuntimeError(`cannot call to_entries against ${input}`);
		return Object.keys(input).map((key) => {
			return {
				key,
				value: input[key]
			};
		});
	}
	exports.toEntries = toEntries;
	function fromEntries(input) {
		if (!Array.isArray(input)) throw new error_1.RuntimeError(`cannot call from_entries against ${input}`);
		return input.reduce((acc, obj) => {
			if (context_1.isJsonMap(obj) && "key" in obj && typeof obj.key === "string" && "value" in obj) {
				const { key, value } = obj;
				return Object.assign(Object.assign({}, acc), { [key]: value });
			}
			throw new error_1.RuntimeError(`cannot call from_entries against ${input}`);
		}, {});
	}
	exports.fromEntries = fromEntries;
	function isArray(input) {
		return Array.isArray(input);
	}
	exports.isArray = isArray;
	function isObject(input) {
		return context_1.isJsonMap(input);
	}
	exports.isObject = isObject;
	function isIterable(input) {
		return isArray(input) || context_1.isJsonMap(input);
	}
	exports.isIterable = isIterable;
	function isBoolean(input) {
		return input === true || input === false;
	}
	exports.isBoolean = isBoolean;
	function isNumber(input) {
		return typeof input === "number";
	}
	exports.isNumber = isNumber;
	function isNormal(input) {
		return typeof input === "number" && isFinite(input) && !isNaN(input) && input != 0;
	}
	exports.isNormal = isNormal;
	function isString(input) {
		return typeof input === "string";
	}
	exports.isString = isString;
	function join(input) {
		if (!isArray(input)) throw new error_1.RuntimeError(`cannot add non-array elements: ${input}`);
		if (input.length === 0) return null;
		return input.reduce((acc, v) => acc ? operators_1.addOperator(acc, v) : v, void 0);
	}
	exports.join = join;
	function anyOf(input) {
		if (!isArray(input)) throw new error_1.RuntimeError(`cannot check any against non-array elements: ${input}`);
		if (input.length === 0) return false;
		return input.some((v) => !utility_1.isFalsy(v));
	}
	exports.anyOf = anyOf;
	function allOf(input) {
		if (!isArray(input)) throw new error_1.RuntimeError(`cannot check all against non-array elements: ${input}`);
		if (input.length === 0) return true;
		return input.every((v) => !utility_1.isFalsy(v));
	}
	exports.allOf = allOf;
	function flattenRecursive(input, depth) {
		if (depth === 0) return input;
		return input.reduce((acc, v) => [...acc, ...isArray(v) ? flattenRecursive(v, depth ? depth - 1 : void 0) : [v]], []);
	}
	function flattenArray(input, depth) {
		if (!isArray(input)) throw new error_1.RuntimeError(`cannot flatten non-array: ${input}`);
		if (depth !== void 0 && typeof depth !== "number") throw new error_1.RuntimeError(`cannot call flatten(depth) with non-number depth: ${JSON.stringify(depth)}`);
		return flattenRecursive(input, depth);
	}
	exports.flattenArray = flattenArray;
	function range(fromOrUpto, uptoOrUndef, byOrUndef) {
		const [from, upto] = uptoOrUndef === void 0 ? [0, fromOrUpto] : [fromOrUpto, uptoOrUndef];
		const by = byOrUndef !== null && byOrUndef !== void 0 ? byOrUndef : 1;
		const order = from < upto ? "asc" : "desc";
		if (order === "asc" && by <= 0) return [];
		if (order === "desc" && by >= 0) return [];
		const values = [];
		for (let i = from; order === "asc" ? i < upto : i > upto; i += by) values.push(i);
		return values;
	}
	exports.range = range;
	function toNumber(input) {
		if (typeof input === "number") return input;
		if (typeof input === "string") return parseFloat(input);
		throw new error_1.RuntimeError(`cannot convert to number from non-number/non-string value: ${input}`);
	}
	exports.toNumber = toNumber;
	function toString(input) {
		if (typeof input === "string") return input;
		else return JSON.stringify(input);
	}
	exports.toString = toString;
	function _type(input) {
		if (input === null) return "null";
		else if (isArray(input)) return "array";
		else return typeof input;
	}
	exports._type = _type;
	function isInfinite(input) {
		return isNumber(input) && !isFinite(input);
	}
	exports.isInfinite = isInfinite;
	function sort(values) {
		return values.sort(utility_1.valueCompare);
	}
	exports.sort = sort;
	function sortBy(values, keys2) {
		return keys2.map((key, index2) => ({
			key,
			index: index2
		})).sort((a, b) => utility_1.valueCompare(a.key, b.key)).map(({ index: index2 }) => values[index2]);
	}
	exports.sortBy = sortBy;
	function groupBy(values, keys2) {
		return [...keys2.reduce((groups, key, i) => {
			var _a;
			return groups.set(key, [...(_a = groups.get(key)) !== null && _a !== void 0 ? _a : [], values[i]]);
		}, /* @__PURE__ */ new Map()).entries()].sort((a, b) => utility_1.valueCompare(a[0], b[0])).map((x) => x[1]);
	}
	exports.groupBy = groupBy;
	function minIndex(input) {
		if (input.length == 0) throw new Error(`cannot get the minimum element from empty array value: ${input}`);
		return input.reduce((minIndex2, v, i) => utility_1.valueCompare(input[minIndex2], v) > 0 ? i : minIndex2, 0);
	}
	exports.minIndex = minIndex;
	function maxIndex(input) {
		if (input.length == 0) throw new Error(`cannot get the minimum element from empty array value: ${input}`);
		return input.reduce((minIndex2, v, i) => utility_1.valueCompare(input[minIndex2], v) < 0 ? i : minIndex2, 0);
	}
	exports.maxIndex = maxIndex;
	function unique(input) {
		return [...input.reduce((acc, v) => acc.add(v), /* @__PURE__ */ new Set()).values()].sort();
	}
	exports.unique = unique;
	function uniqueBy(values, keys2) {
		return groupBy(values, keys2).map((xs) => xs[0]);
	}
	exports.uniqueBy = uniqueBy;
	function reverse(input) {
		return input.reverse();
	}
	exports.reverse = reverse;
	function contains(input, target) {
		if (isString(input)) {
			if (!isString(target)) throw new error_1.RuntimeError(`a non-string value ${JSON.stringify(target)} cannot be contained in a string`);
			return input.indexOf(target) > 0;
		}
		if (isArray(input)) {
			if (!isArray(target)) throw new error_1.RuntimeError(`a non-array value ${JSON.stringify(target)} cannot be contained in an array`);
			return target.every((x) => input.some((y) => contains(y, x)));
		}
		if (isObject(input)) {
			if (!isObject(target)) throw new error_1.RuntimeError(`a non-object value ${JSON.stringify(target)} cannot be contained in an object`);
			return Object.entries(target).every(([key, value]) => key in input && contains(input[key], value));
		}
		return utility_1.valueCompare(input, target) === 0;
	}
	exports.contains = contains;
	function indicesFromString(_input, _target, maxIndices = 0, reverse2 = false) {
		const input = reverse2 ? _input.split("").reverse().join("") : _input;
		const target = reverse2 ? _target.split("").reverse().join("") : _target;
		const indices2 = [];
		let start = 0;
		while (start < input.length && (maxIndices == 0 || indices2.length < maxIndices)) {
			const idx = input.indexOf(target, start);
			if (idx < 0) break;
			indices2.push(idx);
			start = idx + target.length;
		}
		return reverse2 ? indices2.reverse().map((i) => input.length - i - target.length) : indices2;
	}
	function indicesFromArray(_input, _target, maxIndices = 0, reverse2 = false) {
		const input = reverse2 ? _input.reverse() : _input;
		const target = reverse2 ? _target.reverse() : _target;
		if (target.length == 0) return [];
		const indices2 = [];
		let start = 0;
		while (start < input.length && (maxIndices == 0 || indices2.length < maxIndices)) {
			const [head] = target;
			const idx = input.slice(start).findIndex((v) => utility_1.valueCompare(v, head) === 0);
			if (idx < 0) break;
			start += idx;
			if (target.length > 1) {
				const rest = input.slice(start, start + target.length);
				if (!target.every((v, i) => utility_1.valueCompare(v, rest[i]) === 0)) {
					start += 1;
					continue;
				}
			}
			indices2.push(start);
			start += target.length;
		}
		return reverse2 ? indices2.reverse().map((i) => input.length - i - target.length) : indices2;
	}
	function _indices(input, target, maxIndices = 0, reverse2 = false) {
		if (isString(input)) {
			if (!isString(target)) throw new error_1.RuntimeError(`a non-string value ${JSON.stringify(target)} cannot be contained in a string`);
			return indicesFromString(input, target, maxIndices, reverse2);
		}
		if (isArray(input)) return indicesFromArray(input, isArray(target) ? target : [target], maxIndices, reverse2);
		throw new error_1.RuntimeError(`cannot get indices against non-string/non-array value: ${JSON.stringify(input)}`);
	}
	function indices(input, target) {
		return _indices(input, target);
	}
	exports.indices = indices;
	function index(input, target) {
		var _a;
		return (_a = _indices(input, target, 1).shift()) !== null && _a !== void 0 ? _a : null;
	}
	exports.index = index;
	function rindex(input, target) {
		var _a;
		return (_a = _indices(input, target, 1, true).shift()) !== null && _a !== void 0 ? _a : null;
	}
	exports.rindex = rindex;
	function startswith(input, target) {
		if (!isString(input)) throw new error_1.RuntimeError(`cannot apply startswith against non-string value ${JSON.stringify(input)}`);
		if (!isString(target)) throw new error_1.RuntimeError(`cannot apply startswith with non-string value ${JSON.stringify(target)}`);
		return input.startsWith(target);
	}
	exports.startswith = startswith;
	function endswith(input, target) {
		if (!isString(input)) throw new error_1.RuntimeError(`cannot apply startswith against non-string value ${JSON.stringify(input)}`);
		if (!isString(target)) throw new error_1.RuntimeError(`cannot apply startswith with non-string value ${JSON.stringify(target)}`);
		return input.endsWith(target);
	}
	exports.endswith = endswith;
	function _combinations(...input) {
		return utility_1.expandCombination(input, [], (prev, xs) => xs.map((x) => [...prev, x]));
	}
	function isArrayOfArray(value) {
		return value.every((v) => isArray(v));
	}
	function combinations(input, n) {
		if (n === void 0) {
			if (!isArrayOfArray(input)) throw new error_1.RuntimeError(`cannot get combinations from array of non-arrays: ${JSON.stringify(input)}`);
			return _combinations(...input);
		} else {
			if (!isNumber(n)) throw new error_1.RuntimeError(`unexpected non-number: ${JSON.stringify(n)}`);
			return _combinations(...[...Array(n)].map(() => input));
		}
	}
	exports.combinations = combinations;
});
var require_functions = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.functions = exports.standardFunction = void 0;
	var context_1 = require_context();
	var __1 = require_eval();
	var ensure_1 = require_ensure();
	var error_1 = require_error();
	var utility_1 = require_utility();
	var predefined_1 = require_predefined();
	function invokeFunction(input, args, fn) {
		try {
			const value = fn(ensure_1.ensureValue(input), ...args.map(ensure_1.ensureValue));
			if (value === void 0) return;
			return { value };
		} catch (error) {
			return { error };
		}
	}
	function forEachArgs(args, fn) {
		const combinations = utility_1.expandCombination(args.map((arg) => arg.values), [], (prev, xs) => xs.map((x) => [...prev, x]));
		return (combinations.length === 0 ? [[]] : combinations).map((args2) => fn(args2)).reduce((acc, x) => [...acc, ...x], []);
	}
	function forEachInvocation(ctx, args, fn) {
		return new context_1.Context(...utility_1.flatten(ctx.values.map((input) => forEachArgs(args.map((arg) => arg.evaluate(new context_1.Context(input))), (args2) => fn(input, args2)))));
	}
	function standardFunction(fn) {
		return (ctx, args) => {
			return forEachInvocation(ctx, args, (input, args2) => {
				const result = invokeFunction(input, args2, fn);
				return result === void 0 ? [] : [result];
			});
		};
	}
	exports.standardFunction = standardFunction;
	function standardNumericFunction(name, fn) {
		return standardFunction((input, ...args) => {
			if (!predefined_1.isNumber(input)) throw new error_1.RuntimeError(`cannot apply function ${name} against non-number: ${JSON.stringify(input)}`);
			return fn(input, ...args);
		});
	}
	class StandardFunctionEvaluator {
		constructor(fun, args = []) {
			this.fun = fun;
			this.args = args;
		}
		evaluate(ctx) {
			return forEachInvocation(ctx, this.args, (input, args) => {
				const result = invokeFunction(input, args, this.fun);
				return result === void 0 ? [] : [result];
			});
		}
		dump() {
			return { stdfunc: this.fun };
		}
	}
	function selectFunction(fn) {
		return (ctx, args) => {
			return forEachInvocation(ctx, args, (input, args2) => {
				const result = invokeFunction(input, args2, fn);
				if (result === void 0) return [];
				if ("value" in result) return utility_1.isFalsy(result.value) ? [] : [input];
				else return [result];
			});
		};
	}
	function standardArrayFunction(fn) {
		return standardFunction((input, ...args) => {
			if (!Array.isArray(input)) throw new error_1.RuntimeError(`unexpected non-array input: ${JSON.stringify(input)}`);
			return fn(input, ...args);
		});
	}
	function arrayFunctionWithKeyEvaluation(fn) {
		return (ctx, args) => {
			const compareBy = args[0];
			const values = ctx.values.map((input) => {
				const values2 = ensure_1.ensureArray(input);
				return { value: fn(values2, values2.map((value2) => {
					return ensure_1.ensureValue(compareBy.evaluate(new context_1.Context({ value: value2 })).values[0]);
				})) };
			});
			return new context_1.Context(...values);
		};
	}
	exports.functions = new Map([
		["length", standardFunction(predefined_1.length)],
		["utf8bytelength", standardFunction(predefined_1.utf8bytelength)],
		["keys", standardFunction(predefined_1.keys)],
		["keys_unsorted", standardFunction(predefined_1.keysUnsorted)],
		["has", standardFunction(predefined_1.has)],
		["in", standardFunction(predefined_1._in)],
		["map", (ctx, [arg]) => {
			if (!arg) return new context_1.Context({ error: new error_1.RuntimeError(`missing argment for map`) });
			return new __1.ArrayEvaluator([new __1.PipedEvaluator([new __1.IndexedEvaluator(new __1.IdentityEvaluator(), new __1.SpreadEvaluator()), arg])]).evaluate(ctx);
		}],
		["to_entries", standardFunction(predefined_1.toEntries)],
		["from_entries", standardFunction(predefined_1.fromEntries)],
		["with_entries", (ctx, args) => new __1.PipedEvaluator([
			new StandardFunctionEvaluator(predefined_1.toEntries),
			new __1.IndexedEvaluator(new __1.IdentityEvaluator(), new __1.SpreadEvaluator()),
			...args,
			new StandardFunctionEvaluator(predefined_1.fromEntries)
		]).evaluate(ctx)],
		["select", selectFunction((_, pred) => !utility_1.isFalsy(pred))],
		["arrays", selectFunction(predefined_1.isArray)],
		["objects", selectFunction(predefined_1.isObject)],
		["iterables", selectFunction(predefined_1.isIterable)],
		["booleans", selectFunction(predefined_1.isBoolean)],
		["numbers", selectFunction(predefined_1.isNumber)],
		["normals", selectFunction(predefined_1.isNormal)],
		["finites", selectFunction((input) => predefined_1.isNumber(input) && isFinite(input))],
		["strings", selectFunction(predefined_1.isString)],
		["nulls", selectFunction((input) => input === null)],
		["values", selectFunction((input) => input !== null)],
		["scalars", selectFunction((input) => !predefined_1.isIterable(input))],
		["empty", selectFunction(() => false)],
		["error", standardFunction((_, message) => {
			if (message === void 0) throw new error_1.RuntimeError(`missing argment for map`);
			if (typeof message !== "string") throw new error_1.RuntimeError(`error message is not a string: ${message}`);
			throw new error_1.RuntimeError(message);
		})],
		["add", standardFunction(predefined_1.join)],
		["any", standardFunction(predefined_1.anyOf)],
		["all", standardFunction(predefined_1.allOf)],
		["flatten", standardFunction(predefined_1.flattenArray)],
		["range", (input, args) => forEachInvocation(input, args, (_, args2) => {
			const [fromOrUpto, uptoOrUndef, byOrUndef] = args2.map(ensure_1.ensureNumberValue);
			const [from, upto] = uptoOrUndef === void 0 ? [0, fromOrUpto] : [fromOrUpto, uptoOrUndef];
			const by = byOrUndef !== null && byOrUndef !== void 0 ? byOrUndef : 1;
			return predefined_1.range(from, upto, by).map((value) => ({ value }));
		})],
		["floor", standardNumericFunction("floor", Math.floor)],
		["sqrt", standardNumericFunction("sqrt", Math.sqrt)],
		["tonumber", standardFunction(predefined_1.toNumber)],
		["tostring", standardFunction(predefined_1.toString)],
		["type", standardFunction(predefined_1._type)],
		["infinite", standardFunction(() => Infinity)],
		["nan", standardFunction(() => NaN)],
		["isinfinite", standardFunction(predefined_1.isInfinite)],
		["isnan", standardFunction((input) => predefined_1.isNumber(input) && isNaN(input))],
		["isfinite", standardFunction((input) => predefined_1.isNumber(input) && isFinite(input))],
		["isnormal", standardFunction(predefined_1.isNormal)],
		["sort", standardArrayFunction(predefined_1.sort)],
		["sort_by", arrayFunctionWithKeyEvaluation(predefined_1.sortBy)],
		["group_by", arrayFunctionWithKeyEvaluation(predefined_1.groupBy)],
		["min", standardArrayFunction((input) => input[predefined_1.minIndex(input)])],
		["min_by", arrayFunctionWithKeyEvaluation((values, keys) => values[predefined_1.minIndex(keys)])],
		["max", standardArrayFunction((input) => input[predefined_1.maxIndex(input)])],
		["max_by", arrayFunctionWithKeyEvaluation((values, keys) => values[predefined_1.maxIndex(keys)])],
		["unique", standardArrayFunction(predefined_1.unique)],
		["unique_by", arrayFunctionWithKeyEvaluation(predefined_1.uniqueBy)],
		["reverse", standardArrayFunction(predefined_1.reverse)],
		["contains", standardFunction(predefined_1.contains)],
		["indices", standardFunction(predefined_1.indices)],
		["index", standardFunction(predefined_1.index)],
		["rindex", standardFunction(predefined_1.rindex)],
		["inside", standardFunction((input, target) => predefined_1.contains(target, input))],
		["startswith", standardFunction(predefined_1.startswith)],
		["endswith", standardFunction(predefined_1.endswith)],
		["combinations", (ctx, args) => forEachInvocation(ctx, args, (input, args2) => {
			predefined_1.combinations;
			const [nOrUndef] = args2.map(ensure_1.ensureNumberValue);
			return predefined_1.combinations(ensure_1.ensureArray(input), nOrUndef).map((value) => ({ value }));
		})]
	]);
});
var require_eval = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.evaluate = exports.TryCatchEvaluator = exports.PipedEvaluator = exports.ParallelEvaluator = exports.BinaryOperatorEvaluator = exports.FunctionCallEvaluator = exports.IndexedEvaluator = exports.SpreadEvaluator = exports.SliceEvaluator = exports.KeyEvaluator = exports.ObjectEvaluator = exports.ObjectFieldEvaluator = exports.ArrayEvaluator = exports.IdentityEvaluator = exports.IdEvaluator = exports.PrimitiveEvaluator = void 0;
	var context_1 = require_context();
	var functions_1 = require_functions();
	var ensure_1 = require_ensure();
	var error_1 = require_error();
	var utility_1 = require_utility();
	var operators_1 = require_operators();
	class PrimitiveEvaluator {
		constructor(value) {
			this.value = value;
		}
		evaluate(ctx) {
			return new context_1.Context({ value: this.value });
		}
		dump() {
			return this.value;
		}
	}
	exports.PrimitiveEvaluator = PrimitiveEvaluator;
	class IdEvaluator {
		constructor(name) {
			this.name = name;
		}
		evaluate(ctx) {
			return new context_1.Context({ id: this.name });
		}
		dump() {
			return { id: this.name };
		}
	}
	exports.IdEvaluator = IdEvaluator;
	class IdentityEvaluator {
		evaluate(ctx) {
			return ctx;
		}
		dump() {
			return { identity: 0 };
		}
	}
	exports.IdentityEvaluator = IdentityEvaluator;
	class ArrayEvaluator {
		constructor(elements) {
			this.elements = elements;
		}
		evaluate(ctx) {
			try {
				const xs = ctx.values.map((v) => {
					const ctx0 = new context_1.Context(v);
					return utility_1.flatten(this.elements.map((x) => x.evaluate(ctx0)).filter((r) => r.values.length > 0).map((r) => r.values));
				});
				return new context_1.Context(...xs.map((x) => ({ value: x.map(ensure_1.ensureValue) })));
			} catch (error) {
				return new context_1.Context({ error });
			}
		}
		dump() {
			return { array: this.elements.map((x) => x.dump()) };
		}
	}
	exports.ArrayEvaluator = ArrayEvaluator;
	class ObjectFieldEvaluator {
		constructor(key, value) {
			this.key = key;
			this.value = value;
		}
		evaluate(ctx) {
			const keys = this.key.evaluate(ctx).values.map(ensure_1.ensureObjectKey);
			const values = utility_1.flatten(ctx.values.map((v) => this.value.evaluate(new context_1.Context(v).withTarget(v)).values)).map((v) => {
				try {
					return ensure_1.ensureValue(v);
				} catch (_a) {
					return ensure_1.ensureObjectKey(v);
				}
			});
			return utility_1.expandCombination([keys, values], [], (prev, xs) => {
				return xs.map((x) => [...prev, x]);
			}).map(([key, value]) => ({ [key]: value }));
		}
		dump() {
			var _a;
			return {
				key: this.key.dump(),
				value: (_a = this.value) === null || _a === void 0 ? void 0 : _a.dump()
			};
		}
	}
	exports.ObjectFieldEvaluator = ObjectFieldEvaluator;
	class ObjectEvaluator {
		constructor(fields) {
			this.fields = fields;
		}
		evaluate(ctx) {
			try {
				const xs = utility_1.flatten(ctx.values.map((v) => {
					const ctx0 = new context_1.Context(v);
					return utility_1.expandCombination(this.fields.map((f) => f.evaluate(ctx0)), {}, (prev, vs) => {
						return vs.map((v2) => {
							return Object.assign(Object.assign({}, prev), v2);
						});
					});
				}));
				return new context_1.Context(...xs.map((value) => ({ value })));
			} catch (error) {
				return new context_1.Context({ error });
			}
		}
		dump() {
			return { object: this.fields.map((x) => x.dump()) };
		}
	}
	exports.ObjectEvaluator = ObjectEvaluator;
	function indexedByKey(key, value) {
		var _a, _b;
		if (context_1.isJsonMap(value)) {
			if (typeof key === "string") return (_a = value[key]) !== null && _a !== void 0 ? _a : null;
			throw new error_1.RuntimeError(`cannot index non-object value with string: ${typeof value}`);
		}
		if (Array.isArray(value)) {
			if (typeof key === "number") return (_b = value[key >= 0 ? key : value.length + key]) !== null && _b !== void 0 ? _b : null;
			throw new error_1.RuntimeError(`cannot index non-array value with number: ${typeof value}`);
		}
		throw new error_1.RuntimeError(`cannot index non-object/non-array value: ${JSON.stringify(value)}`);
	}
	class KeyEvaluator {
		constructor(key) {
			this.key = key;
		}
		evaluate(ctx) {
			try {
				const indexedValues = this.key.evaluate(ctx).values.map((key) => indexedByKey(ensure_1.ensureObjectKeyOrArrayIndexable(key), ensure_1.ensureValue(ctx.target)));
				return new context_1.Context(...indexedValues.map((value) => ({ value })));
			} catch (error) {
				return new context_1.Context({ error });
			}
		}
		dump() {
			return { key: this.key };
		}
	}
	exports.KeyEvaluator = KeyEvaluator;
	class SliceEvaluator {
		constructor(first, last) {
			this.first = first;
			this.last = last;
		}
		evaluate(ctx) {
			var _a, _b, _c, _d;
			const sliceable = ensure_1.ensureSlicable(ctx.target);
			const start = (_b = (_a = this.first) === null || _a === void 0 ? void 0 : _a.evaluate(ctx).values.map((v) => ensure_1.ensureArrayIndexable(v, sliceable.length))) !== null && _b !== void 0 ? _b : [0];
			const end = (_d = (_c = this.last) === null || _c === void 0 ? void 0 : _c.evaluate(ctx).values.map((v) => ensure_1.ensureArrayIndexable(v, sliceable.length))) !== null && _d !== void 0 ? _d : [sliceable.length];
			const pairs = utility_1.expandCombination([start, end], [], (prev, xs) => {
				return xs.map((x) => [...prev, x]);
			});
			return new context_1.Context(...pairs.map(([first, last]) => sliceable.slice(first, last)).map((value) => ({ value })));
		}
		dump() {
			if (this.first) if (this.last) return [this.first.dump(), this.last.dump()];
			else return [this.first.dump(), "last"];
			else if (this.last) return ["start", this.last.dump()];
		}
	}
	exports.SliceEvaluator = SliceEvaluator;
	class SpreadEvaluator {
		evaluate(ctx) {
			const target = ensure_1.ensureValue(ctx.target);
			if (context_1.isJsonMap(target)) return new context_1.Context(...Object.keys(target).map((k) => target[k]).map((v) => ({ value: v })));
			else if (Array.isArray(target)) return new context_1.Context(...target.map((v) => ({ value: v })));
			else throw new error_1.RuntimeError(`cannot iterate over non-object/non-array value: ${target}`);
		}
		dump() {
			return { spread: 0 };
		}
	}
	exports.SpreadEvaluator = SpreadEvaluator;
	class IndexedEvaluator {
		constructor(target, indexer) {
			this.target = target;
			this.indexer = indexer;
		}
		evaluate(ctx) {
			try {
				const results = this.target.evaluate(ctx).values.map((target) => {
					return this.indexer.evaluate(ctx.withTarget(target));
				});
				return new context_1.Context(...utility_1.flatten(results.map((c) => c.values)));
			} catch (error) {
				return new context_1.Context({ error });
			}
		}
		dump() {
			return {
				indexing: this.target.dump(),
				indexer: this.indexer.dump()
			};
		}
	}
	exports.IndexedEvaluator = IndexedEvaluator;
	class FunctionCallEvaluator {
		constructor(funcname, args) {
			this.funcname = funcname;
			this.args = args;
		}
		evaluate(ctx) {
			const func = functions_1.functions.get(this.funcname);
			if (func === void 0) return new context_1.Context({ error: new error_1.RuntimeError(`unknown function '${this.funcname}'`) });
			return func(ctx, this.args);
		}
		dump() {
			return { call: { [this.funcname]: this.args.map((e) => e.dump()) } };
		}
	}
	exports.FunctionCallEvaluator = FunctionCallEvaluator;
	var knownOperators = {
		"*": operators_1.multiplyOperator,
		"/": operators_1.divideOperator,
		"%": operators_1.moduloOperator,
		"+": operators_1.addOperator,
		"-": operators_1.minusOperator,
		and: operators_1.andOperator,
		or: operators_1.orOperator,
		"//": operators_1.alternativeOperator,
		">": operators_1.greaterThanOperator,
		">=": operators_1.greaterThanOrEqualToOperator,
		"<": operators_1.lessThanOperator,
		"<=": operators_1.lessThanOrEqualToOperator,
		"!=": operators_1.notEqualOperator,
		"==": operators_1.equalOperator
	};
	class BinaryOperatorEvaluator {
		constructor(op, lhs, rhs) {
			this.op = op;
			this.lhs = lhs;
			this.rhs = rhs;
			if (op in knownOperators) this.func = knownOperators[op];
			else throw new error_1.RuntimeError(`cannot apply binary operator '${op}' with ${typeof lhs} and ${typeof rhs}`);
		}
		evaluate(ctx) {
			return functions_1.standardFunction((_, lhs, rhs) => this.func(lhs, rhs))(ctx, [this.lhs, this.rhs]);
		}
		dump() {
			var _a, _b;
			return { [this.op]: [(_a = this.lhs) === null || _a === void 0 ? void 0 : _a.dump(), (_b = this.rhs) === null || _b === void 0 ? void 0 : _b.dump()] };
		}
	}
	exports.BinaryOperatorEvaluator = BinaryOperatorEvaluator;
	class ParallelEvaluator {
		constructor(evaluators) {
			this.evaluators = evaluators;
		}
		evaluate(ctx) {
			return new context_1.Context(...this.evaluators.reduce((acc, x) => {
				return [...acc, ...x.evaluate(ctx).values];
			}, []));
		}
		dump() {
			return { parallel: this.evaluators.map((e) => e.dump()) };
		}
	}
	exports.ParallelEvaluator = ParallelEvaluator;
	class PipedEvaluator {
		constructor(evaluators) {
			this.evaluators = evaluators;
		}
		evaluate(ctx) {
			return this.evaluators.reduce((ctx2, x) => x.evaluate(ctx2), ctx);
		}
		dump() {
			return { piped: this.evaluators.map((e) => e.dump()) };
		}
	}
	exports.PipedEvaluator = PipedEvaluator;
	class TryCatchEvaluator {
		constructor(evaluator) {
			this.evaluator = evaluator;
		}
		evaluate(ctx) {
			return new context_1.Context(...this.evaluator.evaluate(ctx).values.filter((v) => !("error" in v)));
		}
		dump() {
			return { trycatch: this.evaluator };
		}
	}
	exports.TryCatchEvaluator = TryCatchEvaluator;
	function evaluate(evaluator, json) {
		return evaluator.evaluate(new context_1.Context({ value: json })).values.map((v) => {
			if ("id" in v) return v;
			else if ("error" in v) throw v.error;
			else return v.value;
		});
	}
	exports.evaluate = evaluate;
});
var require_optimize = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.optimize = void 0;
	var stmt_1 = require_stmt();
	var eval_1 = require_eval();
	var expr_1 = require_expr();
	var error_1 = require_error();
	function optimizePrimitiveExpr(expr) {
		return new eval_1.PrimitiveEvaluator(expr.value);
	}
	function optimizeIdExpr(expr) {
		return new eval_1.IdEvaluator(expr.value);
	}
	function optimizeIdentityExpr(_expr) {
		return new eval_1.IdentityEvaluator();
	}
	function optimizeArrayExpr(expr) {
		return new eval_1.ArrayEvaluator(expr.elements.map((x) => optimizeExpr(x)));
	}
	function optimizeObjectExpr(expr) {
		return new eval_1.ObjectEvaluator(expr.fields.map((f) => {
			const keyEval = optimizeExpr(f.key);
			if (!f.value) return new eval_1.ObjectFieldEvaluator(keyEval, new eval_1.KeyEvaluator(keyEval));
			else return new eval_1.ObjectFieldEvaluator(keyEval, optimizeExpr(f.value));
		}));
	}
	function optimizeIndexedExpr(expr) {
		if (expr.indexer instanceof expr_1.SpreadIndexer) return new eval_1.IndexedEvaluator(optimizeExpr(expr.expr), new eval_1.SpreadEvaluator());
		if (expr.indexer instanceof expr_1.RangeIndexer) {
			const first = expr.indexer.first ? optimizeExpr(expr.indexer.first) : null;
			const last = expr.indexer.last ? optimizeExpr(expr.indexer.last) : null;
			return new eval_1.IndexedEvaluator(optimizeExpr(expr.expr), new eval_1.SliceEvaluator(first, last));
		}
		if (expr.indexer instanceof expr_1.KeyIndexer) return new eval_1.IndexedEvaluator(optimizeExpr(expr.expr), new eval_1.KeyEvaluator(optimizeExpr(expr.indexer.key)));
		throw new error_1.RuntimeError(`unimplemented indexer: ${expr}`);
	}
	function optimizeFunctaionCallExpr(expr) {
		return new eval_1.FunctionCallEvaluator(expr.funcname, expr.args.map(optimizeExpr));
	}
	function optimizeBinaryOperatorExpr(expr) {
		return new eval_1.BinaryOperatorEvaluator(expr.op, optimizeExpr(expr.lhs), optimizeExpr(expr.rhs));
	}
	function optimizeParallelExpr(expr) {
		if (expr.exprs.length === 1) return optimizeExpr(expr.exprs[0]);
		return new eval_1.ParallelEvaluator(expr.exprs.map(optimizeExpr));
	}
	function optimizeFilteredExpr(expr) {
		if (expr.exprs.length === 1) return optimizeExpr(expr.exprs[0]);
		return new eval_1.PipedEvaluator(expr.exprs.map(optimizeExpr));
	}
	function optimizeOptionalExpr(expr) {
		return new eval_1.TryCatchEvaluator(optimizeExpr(expr.expr));
	}
	function optimizeExpr(expr) {
		if (expr instanceof expr_1.FilteredExpr) return optimizeFilteredExpr(expr);
		if (expr instanceof expr_1.ParallelExpr) return optimizeParallelExpr(expr);
		if (expr instanceof expr_1.BinaryOperatorExpr) return optimizeBinaryOperatorExpr(expr);
		if (expr instanceof expr_1.IndexedExpr) return optimizeIndexedExpr(expr);
		if (expr instanceof expr_1.FunctionCallExpr) return optimizeFunctaionCallExpr(expr);
		if (expr instanceof expr_1.ObjectExpr) return optimizeObjectExpr(expr);
		if (expr instanceof expr_1.ArrayExpr) return optimizeArrayExpr(expr);
		if (expr instanceof expr_1.IdentityExpr) return optimizeIdentityExpr(expr);
		if (expr instanceof expr_1.IdExpr) return optimizeIdExpr(expr);
		if (expr instanceof expr_1.LiteralExpr || expr instanceof expr_1.StringLiteralExpr) return optimizePrimitiveExpr(expr);
		if (expr instanceof expr_1.OptionalExpr) return optimizeOptionalExpr(expr);
		throw new error_1.RuntimeError(`unsupported expression: ${expr.dump()}`);
	}
	function optimize(stmts) {
		const exprs = stmts.stmts.filter((s) => stmt_1.isExprStatement(s));
		if (exprs.length > 1) throw new error_1.RuntimeError(`too munknown expressions`);
		const [root] = exprs;
		if (!root) throw new error_1.RuntimeError(`no valid expression`);
		return optimizeExpr(root.expr);
	}
	exports.optimize = optimize;
});
var JQ = __commonJS((exports) => {
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.JQ = void 0;
	var scanner_1 = require_scanner();
	var parser_1 = require_parser();
	var optimize_1 = require_optimize();
	var eval_1 = require_eval();
	class JQ {
		constructor(evaluator) {
			this.evaluator = evaluator;
		}
		static compile(pattern) {
			const scanner = new scanner_1.Scanner(pattern);
			return new JQ(optimize_1.optimize(parser_1.parse(scanner)));
		}
		evaluate(value) {
			return eval_1.evaluate(this.evaluator, value);
		}
		dump() {
			return this.evaluator.dump();
		}
	}
	exports.JQ = JQ;
	exports.default = JQ;
})().default;
var ZEngine = class {
	cache = {};
	entries = /* @__PURE__ */ new Map();
	jq_entries = /* @__PURE__ */ new Map();
	constructor(url, filterOptions = []) {
		const ws = new WebSocket(url);
		ws.addEventListener("open", () => {
			console.log("WebSocket connected!");
			console.log(`Applied filters: ${JSON.stringify(filterOptions)}`);
			ws.send(`applyFilters:${JSON.stringify(filterOptions)}`);
		});
		ws.addEventListener("close", () => console.log("WebSocket disconnected!"));
		ws.addEventListener("error", (error) => {
			ws.close();
			console.error(error);
		});
		ws.addEventListener("message", (event) => {
			const data = JSON.parse(event.data);
			this.update(data);
		});
	}
	register(key, callback) {
		if (!this.entries.get(key)) this.entries.set(key, /* @__PURE__ */ new Set());
		this.entries.get(key)?.add(callback);
		return callback;
	}
	unregister(key, callback) {
		this.entries.get(key)?.delete(callback);
		if (this.entries.get(key)?.size === 0) this.entries.delete(key);
	}
	update(newData) {
		for (const key of this.entries.keys()) {
			const oldValue = this.search(key, this.cache);
			const newValue = this.search(key, newData);
			const callbacks = this.entries.get(key);
			if (!callbacks) continue;
			for (const callback of callbacks) if (oldValue !== newValue) callback(oldValue, newValue, newData);
		}
		for (const { pattern, callbacks } of this.jq_entries.values()) {
			const oldValue = pattern.evaluate(this.cache)[0];
			const newValue = pattern.evaluate(newData)[0];
			if (!callbacks || oldValue === newValue) continue;
			for (const callback of callbacks) callback(oldValue, newValue, newData);
		}
		this.cache = newData;
	}
	search(key, obj) {
		const attrs = key.split(".");
		let curr = obj;
		for (const attr of attrs) {
			if (!(curr instanceof Object)) return curr;
			curr = curr[attr];
		}
		return curr;
	}
	register_jq(query, callback) {
		const pattern = JQ.compile(query);
		if (!this.jq_entries.get(query)) this.jq_entries.set(query, {
			pattern,
			callbacks: /* @__PURE__ */ new Set()
		});
		this.jq_entries.get(query)?.callbacks.add(callback);
		return callback;
	}
	unregister_jq(query, callback) {
		this.jq_entries.get(query)?.callbacks.delete(callback);
		if (this.jq_entries.get(query)?.callbacks.size === 0) this.jq_entries.delete(query);
	}
};
//#endregion
export { ZEngine };
